includes/admin/reports/class-wc-report-downloads.php

```diff
@@ -0,0 +1,256 @@
+		}
+
+		switch ( $column_name ) {
+			case 'timestamp' :
+				echo esc_html( $item->timestamp );
+				break;
+
+			case 'product' :
+				if ( empty( $product ) ) {
+					break;
+				}
+
+				edit_post_link( $product->get_formatted_name(), '', '', $product->get_id() );
+
+				break;
+
+			case 'order' :
+				if ( ! empty( $permission ) ) {
+					edit_post_link( '#' . $permission->order_id, '', '', $permission->order_id );
+				}
+			
+				break;
+
+			case 'user' :
+				if ( $item->user_id > 0 ) {
+					$user = get_user_by( 'id', $item->user_id );
+
+					if ( ! empty( $user ) ) {
+						$user_description = $user->data->user_nicename;
+						if ( empty( $user_description ) ) {
+							$user_description = $user->data->user_email;
+						}
+
+						echo '<a href="' . esc_url( get_edit_user_link( $item->user_id ) ) . '">' . esc_html( $user_description ) . '</a>';
+					}
+				}
+
+				break;
+
+			case 'user_ip_address' :
+				echo esc_html( $item->user_ip_address );
+
+				break;
+		}
+	}
+
+	/**
+	 * Get columns.
+	 *
+	 * @return array
+	 */
+	public function get_columns() {
+
+		$columns = array(
+			'timestamp'			=> __( 'Download time', 'woocommerce' ),
+			'product'			=> __( 'Product', 'woocommerce' ),
+			'order'				=> __( 'Order', 'woocommerce' ),
+			'user'				=> __( 'User', 'woocommerce' ),
+			'user_ip_address'	=> __( 'User IP Address', 'woocommerce' ),
+		);
+
+		return $columns;
+	}
+
+	/**
+	 * Prepare download list items.
+	 */
+	public function prepare_items() {
+
+		$this->_column_headers = array( $this->get_columns(), array(), $this->get_sortable_columns() );
+		$current_page          = absint( $this->get_pagenum() );
+		// Allow filtering per_page value, but ensure it's at least 1.
+		$per_page              = max( 1, apply_filters( 'woocommerce_admin_downloads_report_downloads_per_page', 20 ) );
+
+		$this->get_items( $current_page, $per_page );
+
+		/**
+		 * Pagination.
+		 */
+		$this->set_pagination_args( array(
+			'total_items' => $this->max_items,
+			'per_page'    => $per_page,
+			'total_pages' => ceil( $this->max_items / $per_page ),
+		) );
+	}
+
+	/**
+	 * No items found text.
+	 */
+	public function no_items() {
+		_e( 'No downloads found.', 'woocommerce' );
+	}
+
+	/**
+	 * Get downloads matching criteria.
+	 *
+	 * @param int $current_page
+	 * @param int $per_page
+	 */
+	public function get_items( $current_page, $per_page ) {
+		global $wpdb;
+
+		$this->max_items = 0;
+		$this->items     = array();
+
+		// Get downloads from database.
+		$table = $wpdb->prefix . WC_Customer_Download_Log_Data_Store::get_table_name();
+
+		$query_from = "FROM {$table} as downloads WHERE 1=1 ";
+
+		// Apply filter by permission.
+		if ( isset( $_GET['permission_id'] ) && is_numeric( $_GET['permission_id'] ) && $_GET['permission_id'] > 0 ) {
+
+			// Ensure the permission and product exist.
+			try {
+				$permission	= new WC_Customer_Download( $_GET['permission_id'] );
```

Comment:
To be on the safe side, we should assign `$_GET['permission_id']` to a variable and cast using `intval` or `absint`.